src/plugindata.h

```diff
@@ -263,8 +265,66 @@ void plugin_configure_single(GtkWidget *parent);
 void plugin_help(void);
 void plugin_cleanup(void);
 
+/** Called by Geany when a plugin library is loaded.
+ *
+ * This is the original entry point. Implement and export this function to be loadable at all.
+ * Then fill in GeanyPlugin::info and GeanyPlugin::funcs of the passed @p plugin. Finally
+ * GEANY_PLUGIN_REGISTER() and specify a minimum supported API version.
+ *
+ * For all glory details please read @ref howto.
+ *
+ * @param plugin The unique plugin handle to your plugin. You must set some fields here.
+ * @param module The GModule associated with your plugin library.
+ * @param geany_api_version The API version exposed by the Geany instance that loads your plugin.
+ *           This probably not the same as the one your plugin was compiled against.
+ *
+ * @since 1.26 (API 225)
+ * @see @ref howto
+ */
+void geany_load_module(GeanyPlugin *plugin, GModule *module, gint geany_api_version);
+
 #endif
 
+/** Callback functions that need to be implemented for every plugin.
+ *
+ * These callbacks should be registered by the plugin within Geany's call to
+ * geany_load_module() by calling geany_plugin_register() with an instance of this type.
+ *
+ * Geany will then call the callbacks at appropriate times. Each gets passed the
+ * plugin-defined data pointer as well as the corresponding GeanyPlugin instance
+ * pointer.
+ *
+ * @since 1.26 (API 225)
+ * @see @ref howto
+ **/
+struct GeanyPluginFuncs
+{
+	/** Array of plugin-provided signal handlers @see PluginCallback */
+	PluginCallback *callbacks;
+	/** Called to initialize the plugin, when the user activates it (must not be @c NULL) */
+	gboolean    (*init)      (GeanyPlugin *plugin, gpointer pdata);
+	/** plugins configure dialog, optional (can be @c NULL) */
+	GtkWidget*  (*configure) (GeanyPlugin *plugin, GtkDialog *dialog, gpointer pdata);
+	/** Called when the plugin should show some help, optional (can be @c NULL) */
+	void        (*help)      (GeanyPlugin *plugin, gpointer pdata);
+	/** Called when the plugin is disabled or when Geany exits (must not be @c NULL) */
```

Comment:
>  @codebrainz, have you an idea how to formulate this so it'd be clear?

The thing that confused me was that I had to call `geany_plugin_set_data()` from inside the `geany_load_module()` function in order to access it as a parameter of the `init()` hook, which I assumed was the "correct" way (otherwise why would that parameter exist?), but I did not realize that it was intended that the data/free_func() was tied to the plugin's module lifetime and not it's init()/cleanup() activation lifetime (which is what I obviously expected).

So I guess adding a blurb explaining that the `data` parameter to `init()` is unlikely to be useful unless you split the `data`'s construction into two parts (allocate only in `geany_load_module()` and initialize it the `init()` hook) would've helped. I think it might also be worth making it really clear that `free_func` won't be called necessarily at `cleanup()` time, but also at some other times (including when the plugin is being _activated_ in the Plugin Manager dialog, which is what made it hard to debug ... why would free_func() be called _before_ my plugin is even activated!?).